src/harness/testcases/heartbeat_testcase.py

```diff
@@ -78,6 +78,198 @@ def test_10_9_4_1_1_1(self):
+    """
+    # Register three devices
+    device_a = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_a.json')))
+    self._sas_admin.InjectFccId({'fccId': device_a['fccId']})
+    device_b = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_b.json')))
+    self._sas_admin.InjectFccId({'fccId': device_a['fccId']})
+    device_c = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_c.json')))
+    self._sas_admin.InjectFccId({'fccId': device_a['fccId']})
+    request = {'registrationRequest': [device_a, device_b, device_c]}
+    response = self._sas.Registration(request)['registrationResponse']
+    cbsd_ids = []
+    for resp in response:
+      self.assertEqual(resp['response']['responseCode'], 0)
+      cbsd_ids.append(resp['cbsdId'])
+    del request, response
+
+    # Create and send grant requests
+    grant_0 = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_0['cbsdId'] = cbsd_ids[0]
+    grant_1 = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_1['cbsdId'] = cbsd_ids[1]
+    grant_2 = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_2['cbsdId'] = cbsd_ids[2]
+    request = {'grantRequest': [grant_0, grant_1, grant_2]}
+    # Check grant response
+    response = self._sas.Grant(request)['grantResponse']
+    self.assertEqual(len(response), 3)
+    grant_ids = []
+    grant_expire_time = []
+    for response_num, resp in enumerate(response):
+      self.assertEqual(resp['cbsdId'], cbsd_ids[response_num])
+      self.assertEqual(resp['response']['responseCode'], 0)
+      grant_ids.append(resp['grantId'])
+      grant_expire_time.append(datetime.strptime(resp['grantExpireTime'],
+                                                 '%Y-%m-%dT%H:%M:%SZ'))
+    del request, response
+
+    # Heartbeat the devices
+    heartbeat_request = []
+    for cbsd_id, grant_id in zip(cbsd_ids, grant_ids):
+      heartbeat_request.append({
+          'cbsdId': cbsd_id,
+          'grantId': grant_id,
+          'operationState': 'GRANTED'
+      })
+    request = {'heartbeatRequest': heartbeat_request}
+    response = self._sas.Heartbeat(request)['heartbeatResponse']
+    # Check the heartbeat response
+    self.assertEqual(len(response), 3)
+    for response_num, resp in enumerate(response):
+      self.assertEqual(resp['cbsdId'], cbsd_ids[response_num])
+      self.assertEqual(resp['grantId'], grant_ids[response_num])
+      transmit_expire_time = datetime.strptime(resp['transmitExpireTime'],
+                                               '%Y-%m-%dT%H:%M:%SZ')
+      self.assertLess(datetime.utcnow(), transmit_expire_time)
+      self.assertLessEqual((transmit_expire_time -
+                       datetime.utcnow()).total_seconds(), 240)
+      self.assertLess(transmit_expire_time, grant_expire_time[response_num])
```

Comment:
The transmit expire time and grant expire time times could be the same: "transmitExpireTime shall be not later than grantExpireTime".

assertLessEqual(transmit_expire_time, grant_expire_time[response_num])